Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware.swift

```diff
@@ -50,36 +56,104 @@ struct ChannelReadUpdaterMiddleware: EventMiddleware {
-        guard let currentUserId = session.currentUser?.user.id, currentUserId != message.user.id else {
-            return
+        
+        if let skipReason = unreadCountUpdateSkippingReason(
+            currentUser: currentUser,
+            channelRead: channelRead,
+            message: message
+        ) {
+            return log.debug(
+                "Message \(message.id) does not increment unread counts for channel \(cid): \(skipReason)",
+                subsystems: .webSocket
+            )
         }
-
-        // Try to get the existing channel read for the current user
-        if let read = session.loadChannelRead(cid: cid, userId: currentUserId) {
-            if message.createdAt > read.lastReadAt ?? Date.distantPast {
-                read.unreadMessageCount += 1
-            }
-        } else {
-            log.info(
-                "Can't increase unread count for \(cid) because the `ChannelReadDTO` for " +
-                    "the current user doesn't exist."
+        
+        log.debug(
+            "Message \(message.id) increments unread counts for channel \(cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount += 1
+    }
+    
+    private func decrementUnreadCountIfNeeded(
+        event: MessageDeletedEventDTO,
+        session: DatabaseSession
+    ) {
+        guard let currentUser = session.currentUser else {
+            return log.error("Current user is missing", subsystems: .webSocket)
+        }
+        
+        guard let channelRead = session.loadChannelRead(cid: event.cid, userId: currentUser.user.id) else {
+            return log.error("Channel read is missing", subsystems: .webSocket)
+        }
+        
+        if let skipReason = !event.hardDelete
+            ? "Soft-deleted messages do not affect unread counts"
+            : unreadCountUpdateSkippingReason(
+                currentUser: currentUser,
+                channelRead: channelRead,
+                message: event.message
+            ) {
+            return log.debug(
+                "Message \(event.message.id) does not decrement unread coutns for \(event.message.user.id): \(skipReason)",
+                subsystems: .webSocket
             )
         }
+        
+        log.debug(
+            "Message \(event.message.id) decrements unread counts for channel \(event.cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount = max(0, channelRead.unreadMessageCount - 1)
+    }
+        
+    private func unreadCountUpdateSkippingReason(
+        currentUser: CurrentUserDTO,
+        channelRead: ChannelReadDTO,
+        message: MessagePayload
+    ) -> String? {
+        guard channelRead.channel.mute == nil else {
+            return "Channel is muted"
+        }
+        
+        guard message.user.id != currentUser.user.id else {
+            return "Own messages do not affect unread counts"
+        }
```

Comment:
`guards` should protect against the regular logic flow. In this case, the function name is "getSkippingReason", so we are actually interested in the skipping reason. Using regular `ifs`, in this case, would be much easier to follow the logic. Where right now we need to think backwards with the guard statements. We should try to avoid using guards in this case

Reply:
Done ✅ Also introduced private `UnreadSkippingReason` enum